core/kernel/dt_driver.c

```diff
@@ -266,3 +296,387 @@ TEE_Result dt_driver_probe_device_by_node(const void *fdt, int nodeoffset,
+
+	DMSG("element: %s on node %s", node_name, drv_name);
+
+	memcpy(elt, &elt_temp, sizeof(*elt));
+
+	probe_list_info.added_count++;
+
+	add_to_probe_list(elt);
+
+	/* Hack here for TRACE_DEBUG messages on current probe list elements */
+	if (TRACE_LEVEL >= TRACE_FLOW)
+		print_probe_list(fdt);
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * Add a node to the probe list if a dt_driver matches target compatible.
+ *
+ * If @type is DT_DRIVER_ANY, probe list can hold only 1 driver to probe for
+ * the node. A node may probe several drivers if have a unique driver type.
+ *
+ * Return TEE_SUCCESS if compatible found
+ *	  TEE_ERROR_ITEM_NOT_FOUND if no matching driver
+ *	  TEE_ERROR_OUT_OF_MEMORY if heap is exhausted
+ */
+static TEE_Result add_probe_node_by_compat(const void *fdt, int node,
+					   const char *compat)
+{
+	TEE_Result res = TEE_ERROR_ITEM_NOT_FOUND;
+	const struct dt_driver *dt_drv = NULL;
+	const struct dt_device_match *dm = NULL;
+	uint32_t found_types = 0;
+
+	for_each_dt_driver(dt_drv) {
+		for (dm = dt_drv->match_table; dm && dm->compatible; dm++) {
+			if (strcmp(dm->compatible, compat) == 0) {
+				assert(dt_drv->type < 32);
+
+				res = add_node_to_probe(fdt, node, dt_drv, dm);
+				if (res)
+					return res;
+
+				if (found_types & BIT(dt_drv->type)) {
+					EMSG("Driver %s multi hit on type %u",
+					     dt_drv->name, dt_drv->type);
+					panic();
+				}
+				found_types |= BIT(dt_drv->type);
+
+				/* Matching found for this driver, try next */
+				break;
+			}
+		}
+	}
+
+	return res;
+}
+
+/*
+ * Add the node to the probe list if matching compatible drivers are found.
+ * Follow node's compatible property list ordering to find matching driver.
+ */
+TEE_Result dt_driver_maybe_add_probe_node(const void *fdt, int node)
+{
+	int idx = 0;
+	int len = 0;
+	int count = 0;
+	const char *compat = NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	if (_fdt_get_status(fdt, node) == DT_STATUS_DISABLED)
+		return TEE_SUCCESS;
+
+	count = fdt_stringlist_count(fdt, node, "compatible");
+	if (count < 0)
+		return TEE_SUCCESS;
+
+	for (idx = 0; idx < count; idx++) {
+		compat = fdt_stringlist_get(fdt, node, "compatible", idx, &len);
+		assert(compat && len > 0);
+
+		res = add_probe_node_by_compat(fdt, node, compat);
+
+		/* Stop lookup if something was found */
+		if (res != TEE_ERROR_ITEM_NOT_FOUND)
+			return res;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static void parse_node(const void *fdt, int node)
+{
+	TEE_Result __maybe_unused res = TEE_ERROR_GENERIC;
+	int subnode = 0;
+
+	fdt_for_each_subnode(subnode, fdt, node) {
+		res = dt_driver_maybe_add_probe_node(fdt, subnode);
+		if (res) {
+			EMSG("Failed on node %s with %#"PRIx32,
+			     fdt_get_name(fdt, subnode, NULL), res);
+			panic();
+		}
+
+		/* Rescursively parse the FDT, skipping disabled nodes */
+		if (IS_ENABLED(CFG_DRIVERS_DT_RECURSIVE_PROBE)) {
+			if (_fdt_get_status(fdt, subnode) == DT_STATUS_DISABLED)
+				continue;
+
+			parse_node(fdt, subnode);
```

Comment:
How deep will this go, worst case?

Reply:
as deep as the DT depth. 
You fear stack exhaustion?

Reply:
Not really, but it would be nice to have some guarantee against it.

Reply:
The DT we're parsing is expected secure. It is authenticated and taken care of with the same care as core execution env itself. Verifying DT consistency is a debug  concern, IMO. `assert()` and like would be enough?

> Not really, (...)

Do you have something else in mind?

Reply:
I guess the motivation above is good enough. Please add a comment like that.